Introduce `iconAlignment` for the buttons with icon

piedcipher opened the pull request:
Adds `iconAlignment` property to `ButtonStyleButton` widget.

Fixes #89564

### Example

https://github.com/flutter/flutter/assets/13456345/1b5236c4-5c60-4915-b3c6-0a56c43f8a19

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

piedcipher committed:
Introduce iconAlignment Property in ButtonStyleButton

TahaTesser commented:
https://github.com/flutter/flutter/issues/89564 was in my backlog. Fo future, please provide an update in the issue if you're working on it on an already assigned issue.  Thanks for taking thus.

First, we will need to support aligning icon in other buttons as well not just  in`TextButton` widget
Secondly, using ` bool` parameter would be limiting, instead we should think about aligning the icon in either direction in both LTR and RTL text directions and prefer an `enum` type.

piedcipher commented:
> #89564 was in my backlog. Fo future, please provide an update in the issue if you're working on an already assigned issue. Thanks for taking this.
> 
> Firstly, we will need to support aligning icon in other buttons as well, not just in `TextButton` widget.
> 
> Secondly, using `bool` parameter would be limiting, instead we should think about aligning the icon in either direction in both LTR and RTL text directions and prefer an `enum` type.

Since there was no active pr I took it. But I'll follow the practice by commenting on issue.

Which other buttons? May be we can create a seperate issue for other buttons?!

I thought about enum but wasn't sure. Will implement.

TahaTesser commented:
> Since there was no active pr I took it. But I'll follow the practice by commenting on issue.

It's best practice on file PRs for unassigned issues to avoid multiple people working on the same problem.

> Which other buttons? May be we can create a seperate issue for other buttons?!

This fix should be included for other `ButtonStyleButton` buttons.
https://master-api.flutter.dev/flutter/material/ButtonStyleButton-class.html

The change would consistent for other buttons so you can the include the fix in this PR. 

Feel free to comment in the issue to say you're fixing this for all `ButtonStyleButton` buttons.

piedcipher commented:
tried branch rename but apparently it closes the pr. will stick to same branch name.

piedcipher commented:
Implemented the suggestions @TahaTesser, please have a look.

TahaTesser commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -16,6 +16,21 @@ import 'material.dart';
+/// {@endtemplate}
+enum IconAlignment {
+  /// icon is placed on the left side of the label
+  left,
```
`left` -> `start` 
`right` -> `end` 

For compatibility with LTR and RTL, see https://main-api.flutter.dev/flutter/dart-ui/TextDirection.html

piedcipher replied:
I've a doubt, so we need to check TextDirection from theme you mean to say?

TahaTesser replied:
Check for the current text direction (`final TextDirection textDirection = Directionality.of(context);`).

If the text direction is LTR and `IconAlignment.start` is set, it aligns the icon to the left and if RTL, to the right.
if the text direction is LTR and `IconAlignment.end` is set, it aligns to the to he right, and if RTL, to the left.

piedcipher replied:
Thanks a lot for this explanation, I couldn't understand at first. Appreciated 👍 🙌

piedcipher replied:
Implemented

piedcipher replied:
I was manually checking the textDirection too in code. I misunderstood your points. Reimplemented this suggestion.

TahaTesser commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -16,6 +16,21 @@ import 'material.dart';
 import 'theme_data.dart';
 
+/// {@template flutter.material.ButtonStyleButton.iconAlignment}
+/// Determines alignment of icon i.e. after/before the label in TextButton.icon,
+/// ElevatedButton.icon, OutlinedButton.icon, FilledButton.icon,
+/// FilledButton.tonalIcon.
```
```suggestion
/// Determines alignment of icon i.e. after/before the label in TextButton.icon,
/// ElevatedButton.icon, OutlinedButton.icon, FilledButton.icon,
/// FilledButton.tonalIcon.
```
Elaborate "Determines alignment of icon i.e. after/before the label" with more details and it should probably say something similar to "start of the button" and "end of the button"

Supported widgets could be in a bullet list

piedcipher replied:
Sure

piedcipher replied:
Will do it today

piedcipher replied:
was stuck on an unexpected mistake - https://github.com/flutter/flutter/pull/137348#discussion_r1436621698

so code-comment will take it up today.

piedcipher replied:
Done

TahaTesser commented on packages/flutter/test/material/elevated_button_test.dart:
```diff
@@ -1914,6 +1914,52 @@ void main() {
     expect(count, 1);
   });
+
+  testWidgetsWithLeakTracking('ElevatedButton.icon honors iconAlignment', (WidgetTester tester) async {
```
```suggestion
  testWidgetsWithLeakTracking('ElevatedButton.icon honors iconAlignment', (WidgetTester tester) async {
```
Split this test into two,  "Default iconAlignment" and "iconAlignment can be customized" (it should  test both start and end position, in both text directions). 

Same for other test classes

piedcipher replied:
I'm confused about the textDirection still.

piedcipher replied:
Implemented

piedcipher replied:
Reimplemented

TahaTesser commented on packages/flutter/test/material/elevated_button_test.dart:
```diff
@@ -1914,6 +1914,52 @@ void main() {
+        ),
+      ),
+    );
+    final Row rowWidget = tester.widget<Row>(
+      find.descendant(
+        of: find.byKey(key),
+        matching: find.byType(Row),
+      ),
+    );
+    expect(rowWidget.children.first.runtimeType, Icon);
+    expect(rowWidget.children.last.runtimeType, Flexible);
```
```suggestion
    final Row rowWidget = tester.widget<Row>(
      find.descendant(
        of: find.byKey(key),
        matching: find.byType(Row),
      ),
    );
    expect(rowWidget.children.first.runtimeType, Icon);
    expect(rowWidget.children.last.runtimeType, Flexible);
```

Rather than testing `runtimeType` in a Row. This should check for the icon itself and its position in the button.

piedcipher replied:
Actually list contains 3 elements: Icon/Text, Spacing, Text/Icon. Hence I used runtimeType approach.

piedcipher replied:
Will do it today

piedcipher replied:
was stuck on an unexpected mistake - https://github.com/flutter/flutter/pull/137348#discussion_r1436621698

will work on correct expect statements today.

piedcipher replied:
last action item on my plate, will wrap it today. sorry for the delay.

piedcipher replied:
Implemented

TahaTesser commented:
@piedcipher 
Do you need more time for this PR?  I see this is still using `left` and `right`

piedcipher commented:
> @piedcipher Do you need more time for this PR? I see this is still using `left` and `right`

Hi @TahaTesser, sorry for the delay. I picked it up last week. I'll get this completed this week. Thanks!

piedcipher commented on packages/flutter/test/material/elevated_button_test.dart:
> Check the gap between the icon and the label

CI is failing because of new changes. Line 1170 here (other similar tests in other buttons).. checking

piedcipher replied:
Fixed.

piedcipher commented:
finally I'm done with all the changes!!

thanks a lot for the suggestions @TahaTesser, please have a look! sorry for the delay

TahaTesser commented on packages/flutter/test/material/elevated_button_test.dart:
```diff
@@ -1944,6 +1944,116 @@ void main() {
+    final Key key = UniqueKey();
+
+    Widget buildFrame({
+      required TextDirection textDirection,
+    }) {
+      return MaterialApp(
+        home: Directionality(
+          textDirection: textDirection,
+          child: Center(
+            child: ElevatedButton.icon(
+              key: key,
+              icon: const Icon(Icons.add),
+              onPressed: () {},
+              label: const Text('button'),
+            ),
+          ),
+        ),
+      );
+    }
+
+    await tester.pumpWidget(
+      buildFrame(
+        textDirection: TextDirection.ltr,
+      ),
+    );
+    Offset iconOffset = tester.renderObject<RenderBox>(find.byIcon(Icons.add)).localToGlobal(Offset.zero);
+    Offset textOffset = tester.renderObject<RenderBox>(find.text('button')).localToGlobal(Offset.zero);
+    Offset buttonOffset = tester.renderObject<RenderBox>(find.descendant(of: find.byKey(key), matching: find.byType(Material))).localToGlobal(Offset.zero);
+    expect(iconOffset.dx - buttonOffset.dx < textOffset.dx - buttonOffset.dx, isTrue);
+
+    await tester.pumpWidget(
+      buildFrame(
+        textDirection: TextDirection.rtl,
+      ),
+    );
+    iconOffset = tester.renderObject<RenderBox>(find.byIcon(Icons.add)).localToGlobal(Offset.zero);
+    textOffset = tester.renderObject<RenderBox>(find.text('button')).localToGlobal(Offset.zero);
+    buttonOffset = tester.renderObject<RenderBox>(find.descendant(of: find.byKey(key), matching: find.byType(Material))).localToGlobal(Offset.zero);
+    expect(textOffset.dx - buttonOffset.dx < iconOffset.dx - buttonOffset.dx, isTrue);
+  });
```
```suggestion
  testWidgets('Default iconAlignment', (WidgetTester tester) async {
    Widget buildWidget({ required TextDirection textDirection }) {
      return MaterialApp(
        home: Directionality(
          textDirection: textDirection,
          child: Center(
            child: ElevatedButton.icon(
              onPressed: () {},
              icon: const Icon(Icons.add),
              label: const Text('button'),
            ),
          ),
        ),
      );
    }

    // Test default iconAlignment when textDirection is ltr.
    await tester.pumpWidget(buildWidget(textDirection: TextDirection.ltr));

    final Offset buttonTopLeft = tester.getTopLeft(find.byType(Material).last);
    final Offset iconTopOffset = tester.getTopLeft(find.byIcon(Icons.add));

    // The icon is aligned to the left of the button.
    expect(buttonTopLeft.dx, iconTopOffset.dx - 16.0); // 16.0 - padding between icon and button edge.

    // Test default iconAlignment when textDirection is rtl.
    await tester.pumpWidget(buildWidget(textDirection: TextDirection.rtl));

    final Offset buttonTopRight = tester.getTopRight(find.byType(Material).last);
    final Offset iconTopRight = tester.getTopRight(find.byIcon(Icons.add));

    // The icon is aligned to the right of the button.
    expect(buttonTopRight.dx, iconTopRight.dx + 16.0); // 16.0 - padding between icon and button edge.
  });
```

TahaTesser commented on packages/flutter/test/material/elevated_button_test.dart:
```diff
@@ -1944,6 +1944,116 @@ void main() {
+      ),
+    );
+    Offset iconOffset = tester.renderObject<RenderBox>(find.byIcon(Icons.add)).localToGlobal(Offset.zero);
+    Offset textOffset = tester.renderObject<RenderBox>(find.text('button')).localToGlobal(Offset.zero);
+    Offset buttonOffset = tester.renderObject<RenderBox>(find.descendant(of: find.byKey(key), matching: find.byType(Material))).localToGlobal(Offset.zero);
+    expect(iconOffset.dx - buttonOffset.dx < textOffset.dx - buttonOffset.dx, isTrue);
+
+    await tester.pumpWidget(
+      buildFrame(
+        iconAlignment: IconAlignment.start,
+        textDirection: TextDirection.rtl,
+      ),
+    );
+    iconOffset = tester.renderObject<RenderBox>(find.byIcon(Icons.add)).localToGlobal(Offset.zero);
+    textOffset = tester.renderObject<RenderBox>(find.text('button')).localToGlobal(Offset.zero);
+    buttonOffset = tester.renderObject<RenderBox>(find.descendant(of: find.byKey(key), matching: find.byType(Material))).localToGlobal(Offset.zero);
+    expect(textOffset.dx - buttonOffset.dx < iconOffset.dx - buttonOffset.dx, isTrue);
+
+    await tester.pumpWidget(
+      buildFrame(
+        iconAlignment: IconAlignment.end,
+        textDirection: TextDirection.ltr,
+      ),
+    );
+    iconOffset = tester.renderObject<RenderBox>(find.byIcon(Icons.add)).localToGlobal(Offset.zero);
+    textOffset = tester.renderObject<RenderBox>(find.text('button')).localToGlobal(Offset.zero);
+    buttonOffset = tester.renderObject<RenderBox>(find.descendant(of: find.byKey(key), matching: find.byType(Material))).localToGlobal(Offset.zero);
+    expect(textOffset.dx - buttonOffset.dx < iconOffset.dx - buttonOffset.dx, isTrue);
+
+    await tester.pumpWidget(
+      buildFrame(
+        iconAlignment: IconAlignment.end,
+        textDirection: TextDirection.rtl,
+      ),
+    );
+    iconOffset = tester.renderObject<RenderBox>(find.byIcon(Icons.add)).localToGlobal(Offset.zero);
+    textOffset = tester.renderObject<RenderBox>(find.text('button')).localToGlobal(Offset.zero);
+    buttonOffset = tester.renderObject<RenderBox>(find.descendant(of: find.byKey(key), matching: find.byType(Material))).localToGlobal(Offset.zero);
+    expect(iconOffset.dx - buttonOffset.dx < textOffset.dx - buttonOffset.dx, isTrue);
+  });
```
```suggestion
  testWidgets('iconAlignment can be customized', (WidgetTester tester) async {
    Widget buildWidget({
      required TextDirection textDirection,
      required IconAlignment iconAlignment,
    }) {
      return MaterialApp(
        home: Directionality(
          textDirection: textDirection,
          child: Center(
            child: ElevatedButton.icon(
              onPressed: () {},
              icon: const Icon(Icons.add),
              label: const Text('button'),
              iconAlignment: iconAlignment,
            ),
          ),
        ),
      );
    }

    // When the layout is ltr and iconAlignment is start, the icon is aligned to the left of the button.
    await tester.pumpWidget(buildWidget(textDirection: TextDirection.ltr, iconAlignment: IconAlignment.start));

    Offset buttonTopLeft = tester.getTopLeft(find.byType(Material).last);
    Offset iconTopOffset = tester.getTopLeft(find.byIcon(Icons.add));

    // The icon is aligned to the left of the button.
    expect(buttonTopLeft.dx, iconTopOffset.dx - 16.0); // 16.0 - padding between icon and button edge.

    // When the layout is ltr and iconAlignment is end, the icon is aligned to the right of the button.
    await tester.pumpWidget(buildWidget(textDirection: TextDirection.ltr, iconAlignment: IconAlignment.end));

    Offset buttonTopRight = tester.getTopRight(find.byType(Material).last);
    Offset iconTopRight = tester.getTopRight(find.byIcon(Icons.add));

    // The icon is aligned to the right of the button.
    expect(buttonTopRight.dx, iconTopRight.dx + 24.0); // 24.0 - padding between icon and button edge.

    // When the layout is rtl and iconAlignment is start, the icon is aligned to the right of the button.
    await tester.pumpWidget(buildWidget(textDirection: TextDirection.rtl, iconAlignment: IconAlignment.start));

    buttonTopRight = tester.getTopRight(find.byType(Material).last);
    iconTopRight = tester.getTopRight(find.byIcon(Icons.add));

    // The icon is aligned to the right of the button.
    expect(buttonTopRight.dx, iconTopRight.dx + 16.0); // 16.0 - padding between icon and button edge.

    // When the layout is rtl and iconAlignment is end, the icon is aligned to the left of the button.
    await tester.pumpWidget(buildWidget(textDirection: TextDirection.rtl, iconAlignment: IconAlignment.end));

    buttonTopLeft = tester.getTopLeft(find.byType(Material).last);
    iconTopOffset = tester.getTopLeft(find.byIcon(Icons.add));

    // The icon is aligned to the left of the button.
    expect(buttonTopLeft.dx, iconTopOffset.dx - 24.0); // 24.0 - padding between icon and button edge.
  });
```

TahaTesser reviewed (changes requested):
Simplified default and custom icon alignments tests in `elevated_button_test.dart`. Please repeat these tests for other button classes.

TahaTesser commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -11,11 +11,43 @@ import 'package:flutter/widgets.dart';
+///   - [FilledButton.icon],
+///   - [FilledButton.tonalIcon].
+///
+/// It is compatible with [TextDirection].
+///   - If textDirection is [TextDirection.ltr] & iconAlignment is [IconAlignment.start],
+///     it aligns the icon at the start of the button.
+///   - If textDirection is [TextDirection.rtl] & iconAlignment is [IconAlignment.start],
+///     it aligns the icon at the end of the button.
+///   - If textDirection is [TextDirection.ltr] & iconAlignment is [IconAlignment.end],
+///     it aligns the icon at the end of the button.
+///   - If textDirection is [TextDirection.rtl] & iconAlignment is [IconAlignment.end],
+///     it aligns the icon at the start of the button.
```
I'm not sure if this is the best way to document these conditions. Hans can suggest better structure here.

HansMuller replied:
Previous comment is a suggestion for the text.

piedcipher replied:
I've applied the suggestion via GitHub. Thank you!

HansMuller commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -11,11 +11,43 @@ import 'package:flutter/widgets.dart';
+///   - [FilledButton.icon],
+///   - [FilledButton.tonalIcon].
+///
+/// It is compatible with [TextDirection].
```
```suggestion
The effect of `iconAlignment` depends on [TextDirection]. If textDirection is [TextDirection.ltr] then [IconAlignment.start] and [IconAlignment.end] align the icon on the left or right respectively.  If textDirection is [TextDirection.rtl] then the alignments are reversed.
```

HansMuller commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -11,11 +11,43 @@ import 'package:flutter/widgets.dart';
+///   - [FilledButton.icon],
+///   - [FilledButton.tonalIcon].
+///
+/// The effect of `iconAlignment` depends on [TextDirection]. If textDirection is [TextDirection.ltr] then [IconAlignment.start] and [IconAlignment.end] align the icon on the left or right respectively.  If textDirection is [TextDirection.rtl] then the alignments are reversed.
+///   - If textDirection is [TextDirection.ltr] & iconAlignment is [IconAlignment.start],
+///     it aligns the icon at the start of the button.
+///   - If textDirection is [TextDirection.rtl] & iconAlignment is [IconAlignment.start],
+///     it aligns the icon at the end of the button.
+///   - If textDirection is [TextDirection.ltr] & iconAlignment is [IconAlignment.end],
+///     it aligns the icon at the end of the button.
+///   - If textDirection is [TextDirection.rtl] & iconAlignment is [IconAlignment.end],
+///     it aligns the icon at the start of the button.
```
This is what I was suggesting for the doc change:

```suggestion
/// The effect of `iconAlignment` depends on [TextDirection]. If textDirection is
/// [TextDirection.ltr] then [IconAlignment.start] and [IconAlignment.end] align the
/// icon on the left or right respectively.  If textDirection is [TextDirection.rtl] the
/// the alignments are reversed.
```

piedcipher replied:
my bad, thanks!

TahaTesser commented:
Looks like test classes suggestions are pending.

piedcipher commented:
> Looks like test classes suggestions are pending.

yes I'll finish them today. sorry I forgot convert it to draft.

piedcipher commented:
sorry for more delay, kinda blocked on #141032, I'll find my way.

piedcipher commented:
implemented all the suggestions! please have a look..

piedcipher commented:
@TahaTesser iconAlignment tests for TextButton are failing. They're off by 4 pixels.

I found the reason. 

https://github.com/flutter/flutter/pull/139014

https://github.com/flutter/flutter/commit/075b9f763f6bec8e8646e0acf78e7ea6407230e4

As you can see, a variable called `padding1x` is missing in TextButton. Also test files in the above commit uses 8 in case of TextButton's padding expect statement compare to 12 in case of other button's padding expect statement

TahaTesser commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -11,11 +11,38 @@ import 'package:flutter/widgets.dart';
+/// Defaults to [IconAlignment.start].
+/// {@endtemplate}
+enum IconAlignment {
+  /// icon is placed at the start of the button
+  start,
+
+  /// icon is placed at the end of the button
```
```suggestion
  /// The icon is placed at the start of the button.
  start,

  /// The icon is placed at the end of the button.
```

TahaTesser commented on packages/flutter/lib/src/material/elevated_button.dart:
```diff
@@ -512,18 +517,21 @@ class _ElevatedButtonWithIcon extends ElevatedButton {
 
 class _ElevatedButtonWithIconChild extends StatelessWidget {
-  const _ElevatedButtonWithIconChild({ required this.label, required this.icon });
+  const _ElevatedButtonWithIconChild({ required this.label, required this.icon, required this.iconAlignment });
```
```suggestion
  const _ElevatedButtonWithIconChild({
    required this.label,
    required this.icon,
    required this.iconAlignment,
  });
```

This is exceeding the `80` char limit.

TahaTesser commented on packages/flutter/lib/src/material/filled_button.dart:
```diff
@@ -524,10 +532,11 @@ class _FilledButtonWithIcon extends FilledButton {
 
 class _FilledButtonWithIconChild extends StatelessWidget {
-  const _FilledButtonWithIconChild({ required this.label, required this.icon });
+  const _FilledButtonWithIconChild({ required this.label, required this.icon, required this.iconAlignment });
```
```suggestion
  const _FilledButtonWithIconChild({
    required this.label,
    required this.icon,
    required this.iconAlignment,
  });
```

This is exceeding the `80` char limit.

TahaTesser commented on packages/flutter/lib/src/material/filled_button.dart:
```diff
@@ -473,10 +479,11 @@ class _FilledButtonWithIcon extends FilledButton {
          autofocus: autofocus ?? false,
          clipBehavior: clipBehavior ?? Clip.none,
-         child: _FilledButtonWithIconChild(icon: icon, label: label)
+         child: _FilledButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment)
```
```suggestion
         child: _FilledButtonWithIconChild(
           icon: icon,
           label: label,
           iconAlignment: iconAlignment,
         ),
```

TahaTesser commented on packages/flutter/lib/src/material/filled_button.dart:
```diff
@@ -492,10 +499,11 @@ class _FilledButtonWithIcon extends FilledButton {
          autofocus: autofocus ?? false,
          clipBehavior: clipBehavior ?? Clip.none,
-         child: _FilledButtonWithIconChild(icon: icon, label: label)
+         child: _FilledButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment)
```
```suggestion
         child: _FilledButtonWithIconChild(
           icon: icon,
           label: label,
           iconAlignment: iconAlignment,
         ),
```

TahaTesser commented on packages/flutter/lib/src/material/elevated_button.dart:
```diff
@@ -480,10 +484,11 @@ class _ElevatedButtonWithIcon extends ElevatedButton {
          autofocus: autofocus ?? false,
          clipBehavior: clipBehavior ?? Clip.none,
-         child: _ElevatedButtonWithIconChild(icon: icon, label: label),
+         child: _ElevatedButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment),
```
```suggestion
         child: _ElevatedButtonWithIconChild(
           icon: icon,
           label: label,
           iconAlignment: iconAlignment,
         ),
```

TahaTesser commented on packages/flutter/lib/src/material/outlined_button.dart:
```diff
@@ -411,10 +415,11 @@ class _OutlinedButtonWithIcon extends OutlinedButton {
          autofocus: autofocus ?? false,
          clipBehavior: clipBehavior ?? Clip.none,
-         child: _OutlinedButtonWithIconChild(icon: icon, label: label),
+         child: _OutlinedButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment),
```
```suggestion
         child: _OutlinedButtonWithIconChild(
           icon: icon,
           label: label,
           iconAlignment: iconAlignment,
         ),
```

TahaTesser commented on packages/flutter/lib/src/material/text_button.dart:
```diff
@@ -476,10 +480,11 @@ class _TextButtonWithIcon extends TextButton {
          autofocus: autofocus ?? false,
          clipBehavior: clipBehavior ?? Clip.none,
-         child: _TextButtonWithIconChild(icon: icon, label: label),
+         child: _TextButtonWithIconChild(icon: icon, label: label, iconAlignment: iconAlignment),
```
```suggestion
         child: _TextButtonWithIconChild(
           icon: icon,
           label: label,
           iconAlignment: iconAlignment,
         ),
```

TahaTesser commented on packages/flutter/test/material/filled_button_test.dart:
```diff
@@ -2016,6 +2016,189 @@ void main() {
+    expect(buttonTopRight.dx, iconTopRight.dx + 16.0); // 16.0 - padding between icon and button edge.
+  });
+
+  testWidgets('(tonalIcon) iconAlignment can be customized', (WidgetTester tester) async {
```
```suggestion
  testWidgets('Tonal icon iconAlignment can be customized', (WidgetTester tester) async {
```

TahaTesser commented on packages/flutter/test/material/filled_button_test.dart:
```diff
@@ -2016,6 +2016,189 @@ void main() {
+    expect(buttonTopLeft.dx, iconTopLeft.dx - 24.0); // 24.0 - padding between icon and button edge.
+  });
+
+  testWidgets('(tonalIcon) Default iconAlignment', (WidgetTester tester) async {
```
```suggestion
  testWidgets('Tonal icon default iconAlignment', (WidgetTester tester) async {
```

TahaTesser reviewed (changes requested):
This is getting there. Minor format improvements and text changes are needed.

I think we should also include an official example on how to use this new `IconAlignment` parameter in the buttons. The example should also contain an option to change text direction when pressing a button.

piedcipher commented:
> This is getting there. Minor format improvements and text changes are needed.
> 
> I think we should also include an official example on how to use this new `IconAlignment` parameter in the buttons. The example should also contain an option to change text direction when pressing a button.

thank a lot for all the suggestions, I've applied. Sure we should include one example, great idea! Let me do that now.

piedcipher commented:
some rebase glitch, it's fixed now.

piedcipher commented:
> This is getting there. Minor format improvements and text changes are needed.
> 
> I think we should also include an official example on how to use this new `IconAlignment` parameter in the buttons. The example should also contain an option to change text direction when pressing a button.

I've added the example for ElevatedButton.icon plus added a test for the same. Also, linked it via dartpad tool annotation. We can replicate the same for others. Please have a look.

TahaTesser commented on examples/api/lib/material/elevated_button/elevated_button.1.dart:
```diff
@@ -0,0 +1,113 @@
+
+import 'package:flutter/material.dart';
+
+/// Flutter code sample for using [ElevatedButton.iconAlignment] parameter.
```
Looks like you chose ElevatedButton for this example. However, `iconAlignment` is also available in all buttons.

piedcipher replied:
Actually wanted replicate the same for other button if elevated one is looking fine.

TahaTesser commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -11,11 +11,38 @@ import 'package:flutter/widgets.dart';
+/// icon on the left or right respectively.  If textDirection is [TextDirection.rtl] the
+/// the alignments are reversed.
+///
+/// Defaults to [IconAlignment.start].
```
1.  Copy the contents and delete `elevated_button.1.dart` and it's test .
2. Let's add `button_style_button` folder in the `api/lib/material` parent folder.
3. Add the copied contents to a new `icon_aligment.0.dart` example class.
4. Update the examples to showcase all the support buttons in a `OverflowBar` 
5. Replace `DropdownMenu` with `ChoiceChip` in a `OverflowBar` for the icon alignment selection and text direction selection.
6.  Add `button_style_button` in the test folder and test the example selections in both text directions,

TahaTesser replied:
Since this doc is already a template, the `icon_aligment.0.dart` example will be available in all button docs. Not just in `ElevatedButton`  docs.

TahaTesser replied:
`DropdownMenu` is a bit much for selection of two items.

`ChoiceChip` should provide easy to access selections

piedcipher replied:
> Since this doc is already a template, the `icon_aligment.0.dart` example will be available in all button docs. Not just in `ElevatedButton` docs.

I'll name it button_style_button.icon_alignment.0.dart as guidelines suggest that. https://github.com/flutter/flutter/tree/master/examples/api#naming

piedcipher replied:
> 1. Copy the contents and delete `elevated_button.1.dart` and it's test .
> 2. Let's add `button_style_button` folder in the `api/lib/material` parent folder.
> 3. Add the copied contents to a new `icon_aligment.0.dart` example class.
> 4. Update the examples to showcase all the support buttons in a `OverflowBar`
> 5. Replace `DropdownMenu` with `ChoiceChip` in a `OverflowBar` for the icon alignment selection and text direction selection.
> 6. Add `button_style_button` in the test folder and test the example selections in both text directions,



> Since this doc is already a template, the `icon_aligment.0.dart` example will be available in all button docs. Not just in `ElevatedButton` docs.



> `DropdownMenu` is bit much for selection of two items.
> 
> `ChoiceChip` should provide easy to access selections.



> > Since this doc is already a template, the `icon_aligment.0.dart` example will be available in all button docs. Not just in `ElevatedButton` docs.
> 
> I'll name it button_style_button.icon_alignment.0.dart as guidelines suggest that. https://github.com/flutter/flutter/tree/master/examples/api#naming

added the new example with all the suggestions, plus added test. For some reason IconAlignment.end test for FilledButton.tonalIcon is failing in `button_style_button.icon_alignment.0_test.dart` (I've commented its code), will check it in morning. thanks a lot for the suggestions!

piedcipher replied:
keeping this one unresolved until I figure the test failure.

TahaTesser replied:
> keeping this one unresolved until I figure the test failure.

When resolving this, can you please merge commits into a single commit as they're too many right now.

piedcipher replied:
Sure I'll squash it to 1.

piedcipher replied:
I'll get this sorted by Monday, still unable to catch what's wrong with that particular test. Sorry for the more delay.

piedcipher replied:
also, linux-analyzer ci check is failing because example is not linked. I'll fix this too.

TahaTesser commented on packages/flutter/lib/src/material/elevated_button.dart:
```diff
@@ -74,13 +74,23 @@ class ElevatedButton extends ButtonStyleButton {
+  ///
+  /// {@macro flutter.material.ButtonStyleButton.iconAlignment}
+  ///
+  /// {@tool dartpad}
+  /// This sample showcases the usage of iconAlignment for ElevatedButton.icon.
+  ///
+  /// ** See code in examples/api/lib/material/elevated_button/elevated_button.1.dart **
```
```suggestion
```

TahaTesser reviewed (changes requested):
This is almost there!

TahaTesser commented on packages/flutter/lib/src/material/elevated_button.dart:
```diff
@@ -74,13 +74,23 @@ class ElevatedButton extends ButtonStyleButton {
+  /// This sample showcases the usage of iconAlignment for ElevatedButton.icon.
+  ///
+  /// ** See code in examples/api/lib/material/elevated_button/elevated_button.1.dart **
+  /// {@end-tool}
+  ///
```
```suggestion
```

Side note: No new line is needed between dartpad tags and code

TahaTesser commented:
This has a bunch of conflicts.

piedcipher commented:
> This has a bunch of conflicts.

Let me resolve em.

piedcipher commented:
Commit Squashing :heavy_check_mark: 
Merge Conflicts :heavy_check_mark: 
Linux Analyzer Failure - Usage of Example :hourglass_flowing_sand: 
Filled Button Tonal Icon Example's Test (Commented) - :hourglass_flowing_sand:

piedcipher committed:
link example - make linux-analyzer happy

piedcipher commented on examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart:
```diff
@@ -0,0 +1,213 @@
+    );
+    // expect(
+    //   tester.widget<FilledButton>(
+    //     find.byKey(const Key('FilledButton.tonalIcon')),
```
Not sure why this is failing. IconAlignment is set to IconAlignment.end & Direction is ltr. But still actuall value is IconAlignment.start. only happens for tonalIcon.

TahaTesser commented on examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart:
```diff
@@ -0,0 +1,213 @@
+    );
+    // expect(
+    //   tester.widget<FilledButton>(
+    //     find.byKey(const Key('FilledButton.tonalIcon')),
```
I'll take a look tomorrow

piedcipher replied:
Sure thanks a lot! I've a backup in my mind. Instead of checking iconAlignment, I'll assert for the text.

```dart
expect(finder, 'ElevatedButton.icon start ltr');
expect(finder, 'ElevatedButton.icon end rtl');
// Etc...
```

TahaTesser committed:
Update code sample and test

piedcipher commented:
Thanks, I'll check it out. Also I'll resolve these conflicts came after #142644.

piedcipher commented:
tests are passing locally, something flaky, doing a rebase.

piedcipher commented:
conflicts are back, hans's pr got reverted! I'll get this fixed in morning!

TahaTesser commented:
Usually complex PRs needs second attempts. It relanded https://github.com/flutter/flutter/pull/142762

HansMuller commented:
Sorry about the button churn. Hopefully this time I'll stick the landing (https://github.com/flutter/flutter/pull/142762).

piedcipher commented:
> Sorry about the button churn. Hopefully this time I'll stick the landing (#142762).

please don't say sorry! <3

piedcipher committed:
added couple of tiny gaps

Update examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Update examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Update examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Update examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Update examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Update examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Update packages/flutter/lib/src/material/button_style_button.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Update examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

piedcipher commented:
finallyy this one is ready for a review.

TahaTesser commented on packages/flutter/test/material/elevated_button_test.dart:
```diff
@@ -2178,6 +2178,98 @@ void main() {
+      );
+    }
+
+    // When the layout is ltr and iconAlignment is start, the icon is aligned to the left of the button.
```
```suggestion
    // Test iconAlignment when textDirection is ltr.
```

"the icon is aligned to the left of the button." is repeated. We can keep this short.

TahaTesser replied:
Similar fix for the tests in this class, as well  as in other classes for this PR.

piedcipher replied:
I'll fix it. 

> // Test iconAlignment when textDirection is ltr.

should I mention iconAlignment as well or can be skipped?

piedcipher replied:
Done, went with original suggestion.

TahaTesser commented on packages/flutter/test/material/elevated_button_test.dart:
```diff
@@ -2178,6 +2178,98 @@ void main() {
+    }
+
+    // When the layout is ltr and iconAlignment is start, the icon is aligned to the left of the button.
+    await tester.pumpWidget(buildWidget(textDirection: TextDirection.ltr, iconAlignment: IconAlignment.start));
```
```suggestion
    await tester.pumpWidget(buildWidget(
      textDirection: TextDirection.ltr,
      iconAlignment: IconAlignment.start,
    ));
```

This exceed the 80 characters line limit. Add line breaks in the parameter to readability and 80 charactesr limit.

TahaTesser replied:
Similar fix for the tests in this class, as well as in other classes for this PR.

HansMuller replied:
Note also: there's no hard 80 character line limit however in cases like this the code is easier to read if you apply the conventional Flutter formatting rules.

piedcipher replied:
Done!

TahaTesser reviewed (changes requested):
This is almost there

piedcipher committed:
formatting

piedcipher committed:
remove dups

piedcipher committed:
smol format

TahaTesser commented on packages/flutter/lib/src/material/button_style_button.dart:
```diff
@@ -11,11 +11,45 @@ import 'package:flutter/widgets.dart';
+/// Defaults to [IconAlignment.start].
+///
+/// {@tool dartpad}
+/// This sample demonstrates the usage of iconAlignment.
```
```suggestion
/// This sample demonstrates how to use `iconAlignment` to align the button icon to the start
// or the end of the button.
```

TahaTesser reviewed (approved):
LGTM!

piedcipher committed:
Update packages/flutter/lib/src/material/button_style_button.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

piedcipher commented:
I suspect something flaky is ci, earlier checks were passing.

piedcipher committed:
fix doc comment

piedcipher commented:
from Luci

```
 00:50 +208: /Volumes/Work/s/w/ir/x/w/flutter/examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart: ButtonStyleButton.iconAlignment updates button icons alignment
 ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
 The following TestFailure was thrown running a test:
 Expected: <321.3000030517578>
   Actual: <534.6999969482422>
 
 When the exception was thrown, this was the stack:
 #4      main.<anonymous closure>.expectedRightIconPosition (file:///Volumes/Work/s/w/ir/x/w/flutter/examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart:52:7)
 #5      main.<anonymous closure> (file:///Volumes/Work/s/w/ir/x/w/flutter/examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart:93:5)
 <asynchronous suspension>
 #6      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:183:15)
 <asynchronous suspension>
 #7      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1019:5)
 <asynchronous suspension>
 <asynchronous suspension>
 (elided one frame from package:stack_trace)
 
 This was caught by the test expectation on the following line:
   file:///Volumes/Work/s/w/ir/x/w/flutter/examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart line 52
 The test description was:
   ButtonStyleButton.iconAlignment updates button icons alignment
```

will debug this tomorrow, earlier this didn't occur.

TahaTesser commented on examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    // Test initial icon alignment in LTR.
+    expectedLeftIconPosition(iconOffset: 16, textButtonIconOffset: 12) ;
```
```suggestion
    expectedLeftIconPosition(iconOffset: 16, textButtonIconOffset: 12);
```

TahaTesser commented on examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart:
```diff
@@ -0,0 +1,102 @@
+    await tester.pumpAndSettle();
+
+    // Test icon alignment start in LTR.
+    expectedRightIconPosition(iconOffset: 16, textButtonIconOffset: 12) ;
```
```suggestion
    expectedRightIconPosition(iconOffset: 16, textButtonIconOffset: 12);
```

TahaTesser commented on examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart:
```diff
@@ -0,0 +1,147 @@
+                        selected: <TextDirection>{ _textDirection },
+                        segments: const <ButtonSegment<TextDirection>>[
+                          ButtonSegment<TextDirection>(
+                            value: TextDirection.ltr,
+                            label: Text('RTL'),
```
```suggestion
                            value: TextDirection.ltr,
                            label: Text('LTR'),
```

piedcipher replied:
OMG! how I couldn't figure out this, sorry for the trouble. thank you!

TahaTesser commented on examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart:
```diff
@@ -0,0 +1,147 @@
+                          ),
+                          ButtonSegment<TextDirection>(
+                            value: TextDirection.rtl,
+                            label: Text('LTR'),
```
```suggestion
                            label: Text('RTL'),
```

TahaTesser reviewed (changes requested):
This should fix the example test.

TahaTesser committed:
Update examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart

TahaTesser committed:
Update examples/api/test/material/button_style_button/button_style_button.icon_alignment.0_test.dart

TahaTesser committed:
Update examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart

TahaTesser committed:
Update examples/api/lib/material/button_style_button/button_style_button.icon_alignment.0.dart

piedcipher commented:
I've updated the video too in the pr desc.

TahaTesser reviewed (approved):
LGTM!

HansMuller reviewed (approved):
LGTM
